File: backend/schemas/book_schema.py
from marshmallow import Schema, fields, validate, ValidationError
from models.book_model import Book

class BookSchema(Schema):
    id = fields.Int(dump_only=True)
    title = fields.Str(required=True, validate=validate.Length(min=1, error="Title cannot be empty"))
    author = fields.Str(required=True, validate=validate.Length(min=1, error="Author cannot be empty"))
    cover = fields.Url(allow_none=True)
    rating = fields.Int(validate=validate.Range(min=0, max=5, error="Rating must be between 0 and 5"))
    pages = fields.Int(allow_none=True)
    genre = fields.Str(allow_none=True)
    status = fields.Str(validate=validate.OneOf(["want-to-read", "reading", "completed"], error="Status must be one of: want-to-read, reading, completed"))
    created_at = fields.DateTime(dump_only=True)

File: frontend/src/schemas/book.schema.ts
import { z } from 'zod';

export const bookSchema = z.object({
  title: z.string().min(1, 'Title is required'),
  author: z.string().min(1, 'Author is required'),
  cover: z.string().url('Must be a valid URL').optional().or(z.literal('')),
  rating: z.coerce.number().min(0, 'Rating must be at least 0').max(5, 'Rating must be at most 5').default(0),
  pages: z.coerce.number().positive('Pages must be a positive number').optional().or(z.literal(0)),
  genre: z.string().optional(),
  status: z.enum(['want-to-read', 'reading', 'completed'], {
    required_error: "Status is required",
    invalid_type_error: "Invalid status",
  }),
});

export type BookFormValues = z.infer<typeof bookSchema>;
